btc/client.go

```diff
@@ -377,81 +373,299 @@ type DescendantTransaction struct {
-		Error  interface{}                      `json:"error"`
-		ID     string                           `json:"id"`
+type RBFTxFeeInfo struct {
+	// Total fees of in-mempool descendants (including this transaction) in sats
+	TotalFee float64 `json:"total_fee"`
+	// Total fee of the descendants in sats
+	DescendantFee float64 `json:"descendant_fee"`
+	// current fee of the tx to be replaced ,
+	// the fee rate is calculated based on the actual tx and its direct descendants
+	// source : https://github.dev/bitcoin/bitcoin/blob/ed060e01e756fc8d4c1c590e79f9555ae86c433f/src/policy/rbf.cpp#L145
+	TxFeeRate float64 `json:"tx_fee_rate"`
+}
+
+func (client *BitcoinClient) GetRBFTxFeeInfo(ctx context.Context, txid string) (*RBFTxFeeInfo, error) {
+	descendants, err := client.GetMempoolDescendants(ctx, txid)
+	if err != nil {
+		if errors.Is(err, ErrTxNotFound) {
+			return nil, ErrTxNotFound
+		}
+		return nil, fmt.Errorf("error getting mempool descendants: %w", err)
+	}
+	entry, err := client.GetMempoolEntry(ctx, txid)
+	if err != nil {
+		if errors.Is(err, ErrTxNotFound) {
+			return nil, ErrTxNotFound
+		}
+		return nil, fmt.Errorf("error getting mempool entry: %w", err)
+	}
+	if entry == nil {
+		return nil, ErrTxNotFound
 	}
 
-	if err := json.Unmarshal(body, &apiResponse); err != nil {
-		return 0, fmt.Errorf("Error unmarshalling JSON: %w", err)
+	// Fees are in BTC, convert to sats
+	totalFee := math.Ceil(entry.Fees.Descendant * 1e8)
+	baseFee := math.Ceil(entry.Fees.Base * 1e8)
+	descendantFee := math.Ceil((entry.Fees.Descendant - entry.Fees.Base) * 1e8)
+
+	feeInfo := &RBFTxFeeInfo{
+		TotalFee:      totalFee,
+		DescendantFee: descendantFee,
+		TxFeeRate:     float64(totalFee) / float64(entry.DescendantSize),
+	}
+	if len(descendants) == 0 {
+		return feeInfo, nil
 	}
 
-	if len(apiResponse.Result) == 0 {
-		return 0, nil
+	var directDescendantSize, directDescendantFee float64
+	for _, childTxid := range entry.SpendBy {
+		if desc, ok := descendants[childTxid]; ok {
+			directDescendantSize += float64(desc.Vsize)
+			directDescendantFee += float64(desc.Fees.Base * 1e8)
+		}
 	}
+	if directDescendantSize+float64(entry.VSize) > 0 {
+		feeInfo.TxFeeRate = float64(directDescendantFee+baseFee) / float64(directDescendantSize+float64(entry.VSize))
+	}
+
+	return feeInfo, nil
+}
 
-	transactions := make([]DescendantTransaction, 0, len(apiResponse.Result))
-	for _, tx := range apiResponse.Result {
-		transactions = append(transactions, tx)
+func (client *BitcoinClient) packParams(params ...interface{}) ([]json.RawMessage, error) {
+	rawParams := make([]json.RawMessage, len(params))
+	for i, param := range params {
+		var err error
+		rawParams[i], err = json.Marshal(param)
+		if err != nil {
+			return nil, err
+		}
 	}
+	return rawParams, nil
+}
 
-	var sum float64
-	for _, tx := range transactions {
-		sum += tx.Fees.Base
+func (client *BitcoinClient) send(ctx context.Context, method string, params []json.RawMessage) ([]byte, error) {
```

Comment:
let's remove retrying here. Whoever is using the BitcoinClient should handle what they want to do.